### DISCUSSION TITLE
Slow Javascript Performance

### DISCUSSION THREAD

**Discussion by ducopro** [livewire/livewire#6115] [7. Bugs (archived)]
### Livewire version

v3 beta5

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

```php
<?php

namespace App\Livewire;

use Livewire\Component;

class Preformance extends Component
{
    public bool $mode = true;

    public function setMode(): void
    {
        $this->mode = !$this->mode;
    }

    public function render()
    {

        return <<<'HTML'
        <div>
            <button wire:click="setMode()" class="m-2 bg-red-50">Click me to toggle: {{$mode ? 'State-1' : 'State-2'}}</button>
            <br />
            <br />
            <br />

            <div class="grid grid-cols-1 gap-6 mt-6">
                @for ($i = 0; $i < 1000; $i++)

                    <!-- __BLO__CK__ -->
                    <div wire:key="{{$i}}">
                            <div>
                                <span>
                                    Counterspan: [{{ $i }}]
                                    <small>
                                        {{$mode ? 'Original State' : 'Toggled State'}}
                                    </small>
                                </span>
                            </div>
                    </div>
                    <!-- __ENDBLO__CK__ -->

                @endfor

            </div>
        </div>
        HTML;
    }
}
```

Allright, look at the livewire component above. According to the docs the main div within the loop has a wire:key. You would expect that the wire:key not only helps with correct dom diffing, but also with performance.

Wrong. As you can see, I disabled the block marker with two extra underscores. When you enable the block marker the performance is a lot better. *A lot!* 

If you want a closer look at the javascript performance:
- Use Google Chrome
- Open developer tools
- Open performance Monitor

Experiment with a bigger number in the loop. Don't make it too big! Start to look at 1000 en 1500. As you also can see the performance degradation is not linear!

Again, using the block markers makes the transition near instant.

### Additional information

The above test was done on a Windows machine with a fresh Laravel 10 and Livewire install. No config items where changed, nothing. 

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment]
@ducopro thanks for reporting! I'm not sure I follow, have you manually added the block markers? If so, they should be added automatically for you but only wrapping the `@foreach`, they shouldn't be wrapping each element inside the loop.

**Reply by ducopro** [reply]
Hi Josh,

I am sorry i might have led you on the wrong track with the blockmarkers. Just forget the block markers for now. The block markers are there as part of my experiments. Yes, **I** put them there. I put them there to prove that the javascript performance **can** be excellent.

So forget the blockmarkers. 

I you run my component on a page (without blockmarkers!!!) and check out the javascript performance with the performance monitor in chrome you will see a very long spike (4-5 seconds at a loop size of 1000 on my machine). At this spike the javascript engine is working at a 100% rate and the page is non responsive during this time.

Make the loop size 2000 and you will be up to about 30 seconds. Non-responsive!!! All the time the javascript engine will be running at 100%. This is a non-lineair process

So I also ran a javascript trace. As you expect all the time is spend in the livewire script method patch(). However more than 85% of that time is used in the next() method trying to find the next div element in the loop.

In my experiments to manipulate the time used in the next() method I first tried a wire:key on the main div in the loop. This had absolutely no effect.

My experiment with a block marker however had a very big effect!! Time used in the next method went to almost zero.

So, my analysis is that Caleb already did some very good work getting the dom-diffing up to speed. The block markers are doing a good job. However, the docs only mention setting a wire:key on a div in a loop. Relying on only a wire key can lead to dramatic performance loss as I try to prove with my example component.

Hope I gave you some more insight this time.

**Reply by joshhanley** [reply]
@ducopro all good! Any chance you can test on the latest beta? I know Caleb has just reworked the morphing algorithm so maybe there was some performance changes there.

**Reply by ducopro** [reply]
@joshhanley thanx for your time. I can confirm that the javascript performance issues are gone in the Beta 8 version.
